src/strings.rs

```diff
@@ -0,0 +1,211 @@
+///
+/// assert!(array.dtype().to_string().contains("S3"));
+/// # });
+/// ```
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PyFixedString<const N: usize>(pub [Py_UCS1; N]);
+
+impl<const N: usize> fmt::Display for PyFixedString<N> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str(str::from_utf8(&self.0).unwrap().trim_end_matches('\0'))
+    }
+}
+
+impl<const N: usize> From<[Py_UCS1; N]> for PyFixedString<N> {
+    fn from(val: [Py_UCS1; N]) -> Self {
+        Self(val)
+    }
+}
+
+unsafe impl<const N: usize> Element for PyFixedString<N> {
+    const IS_COPY: bool = true;
+
+    fn get_dtype(py: Python) -> &PyArrayDescr {
+        static DTYPES: TypeDescriptors = TypeDescriptors::new(NPY_TYPES::NPY_STRING, b'|' as _);
+
+        unsafe { DTYPES.from_size(py, size_of::<Self>()) }
+    }
+}
+
+/// A newtype wrapper around [`[PyUCS4; N]`][Py_UCS4] to satisfy coherence.
+///
+/// Note that when creating arrays of Unicode strings without an explicit `dtype`, e.g.
+///
+/// ```python
+/// numpy.array(["foo", "bar", "foobar"])
+/// ```
+///
+/// yielding `U6`, NumPy will automatically determine the smallest possible array length at runtime.
+/// This does not work well with the length `N` of `PyUSC4Array<N>` determined at compile-time
+/// and hence explicitly setting the `dtype`, e.g.
+///
+/// ```python
+/// numpy.array(["foo", "bar", "foobar"], dtype='U12')
+/// ```
+///
+/// always matching `PyArray1<PyUSC4Array<12>>`, is recommended.
+///
+/// # Example
+///
+/// ```rust
+/// # use pyo3::Python;
+/// use numpy::{PyArray1, PyFixedUnicode};
+///
+/// # Python::with_gil(|py| {
+/// let array = PyArray1::<PyFixedUnicode<3>>::from_vec(py, vec![[b'b' as _, b'a' as _, b'r' as _].into()]);
+///
+/// assert!(array.dtype().to_string().contains("U3"));
+/// # });
+/// ```
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PyFixedUnicode<const N: usize>(pub [Py_UCS4; N]);
+
+impl<const N: usize> fmt::Display for PyFixedUnicode<N> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        for character in self.0 {
+            if character == 0 {
+                break;
+            }
+
+            write!(fmt, "{}", char::from_u32(character).unwrap())?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<const N: usize> From<[Py_UCS4; N]> for PyFixedUnicode<N> {
+    fn from(val: [Py_UCS4; N]) -> Self {
+        Self(val)
+    }
+}
+
+unsafe impl<const N: usize> Element for PyFixedUnicode<N> {
+    const IS_COPY: bool = true;
+
+    fn get_dtype(py: Python) -> &PyArrayDescr {
+        static DTYPES: TypeDescriptors = TypeDescriptors::new(NPY_TYPES::NPY_UNICODE, b'=' as _);
+
+        unsafe { DTYPES.from_size(py, size_of::<Self>()) }
+    }
+}
+
+struct TypeDescriptors {
+    npy_type: NPY_TYPES,
+    byteorder: c_char,
```

Comment:
I'm not sure we should hold these values as static. We can just give these values to `from_size`, and they are converted to constants when compiling.